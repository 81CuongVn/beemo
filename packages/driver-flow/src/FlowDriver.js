/**
 * @copyright   2017, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 * @flow
 */

import { Driver } from '@beemo/core';

import type { Execution } from '@beemo/core';

// Success: Writes no errors message to stdout and server output to stderr
// Failure: Writes file list to stdout and server output to stderr
export default class FlowDriver extends Driver {
  bootstrap() {
    this.setMetadata({
      bin: 'flow',
      configName: '.flowconfig',
      description: 'Type check files with Flow.',
      title: 'Flow',
    });
  }

  formatFile(data: Object): string {
    const output = [];

    Object.keys(data).forEach(key => {
      const value = data[key];

      if (!value) {
        return;
      }

      output.push(`[${key}]`);

      switch (key) {
        default:
          output.push(...value.map(v => String(v)));
          break;
        case 'lints':
          output.push(...this.formatLintsSection(value));
          break;
        case 'options':
          output.push(...this.formatOptionsSection(value));
          break;
        case 'version':
          if (value) {
            output.push(String(value));
          }
          break;
      }

      output.push('');
    });

    return output.join('\n');
  }

  formatLintsSection(lints: Object): string[] {
    const output = [];

    Object.keys(lints).forEach(key => {
      let value = lints[key];

      if (value === 0) {
        value = 'off';
      } else if (value === 1) {
        value = 'warn';
      } else if (value === 2) {
        value = 'error';
      }

      output.push(`${key.replace(/_/g, '-')}=${String(value)}`);
    });

    return output;
  }

  formatOption(value: *, quote: boolean = false): string {
    let option = value;

    // http://caml.inria.fr/pub/docs/manual-ocaml/libref/Str.html#TYPEregexp
    if (value instanceof RegExp) {
      option = value.source
        .replace(/\|/g, '\\|')
        .replace(/\(/g, '\\(')
        .replace(/\)/g, '\\)');
    } else {
      option = String(value);
    }

    return quote ? `'${option}'` : option;
  }

  formatOptionsSection(options: Object): string[] {
    const output = [];

    Object.keys(options).forEach(key => {
      const value = options[key];

      // Multiple values
      if (Array.isArray(value)) {
        value.forEach(val => {
          output.push(`${key}=${this.formatOption(val)}`);
        });

        // Mapped objects
      } else if (typeof value === 'object' && value.constructor === Object) {
        Object.keys(value).forEach(pattern => {
          output.push(
            `${key}=${this.formatOption(pattern, true)} -> ${this.formatOption(
              value[pattern],
              true,
            )}`,
          );
        });

        // Primitives
      } else {
        output.push(`${key}=${this.formatOption(value)}`);
      }
    });

    return output;
  }

  // https://github.com/facebook/flow/blob/e466b0ee519622a8977e89708be156a73e570ef0/hack/utils/exit_status.ml#L78
  // https://github.com/facebook/flow/blob/e466b0ee519622a8977e89708be156a73e570ef0/src/common/flowExitStatus.ml#L54
  handleFailure(error: Execution) {
    if (error.code === 2) {
      this.tool.logError(error.stdout); // Command failures
    } else {
      super.handleFailure(error);
    }
  }
}
