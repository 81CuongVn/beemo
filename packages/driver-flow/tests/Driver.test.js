import fs from 'fs';
import { Event } from 'boost';
import FlowDriver from '../src/FlowDriver';

describe('FlowDriver', () => {
  let driver;

  beforeEach(() => {
    driver = new FlowDriver();
    driver.bootstrap();
  });

  it('sets options from constructor', () => {
    driver = new FlowDriver({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: true },
    });

    expect(driver.options).toEqual({
      args: ['--foo', '--bar=1'],
      dependencies: ['babel'],
      env: { DEV: true },
    });
  });

  it('sets correct metadata', () => {
    expect(driver.metadata).toEqual({
      bin: 'flow',
      configName: '.flowconfig',
      configOption: '--config',
      dependencies: [],
      description: 'Type check files with Flow.',
      helpOption: '--help',
      title: 'Flow',
      useConfigOption: false,
    });
  });

  describe('formatFile()', () => {
    describe('[include]', () => {
      it('supports', () => {
        expect(driver.formatFile({
          include: ['./foo.js', './bar'],
        })).toMatchSnapshot();
      });

      it('handles empty', () => {
        expect(driver.formatFile({
          include: [],
        })).toMatchSnapshot();
      });
    });

    describe('[ignore]', () => {
      it('supports', () => {
        expect(driver.formatFile({
          ignore: ['.*/__tests__/.*'],
        })).toMatchSnapshot();
      });

      it('handles empty', () => {
        expect(driver.formatFile({
          ignore: [],
        })).toMatchSnapshot();
      });
    });

    describe('[libs]', () => {
      it('supports', () => {
        expect(driver.formatFile({
          libs: ['./foo.js.flow'],
        })).toMatchSnapshot();
      });

      it('handles empty', () => {
        expect(driver.formatFile({
          libs: [],
        })).toMatchSnapshot();
      });
    });

    describe('[lints]', () => {
      it('supports', () => {
        expect(driver.formatFile({
          lints: {
            all: 'error',
            untyped_type_import: 'warn',
            sketchy_null_bool: 'off',
          },
        })).toMatchSnapshot();
      });

      it('handles values as numbers', () => {
        expect(driver.formatFile({
          lints: {
            all: 2,
            untyped_type_import: 1,
            sketchy_null_bool: 0,
          },
        })).toMatchSnapshot();
      });

      it('supports dashed keys', () => {
        expect(driver.formatFile({
          lints: {
            all: 'error',
            'untyped-type-import': 'warn',
            'sketchy-null-bool': 'off',
          },
        })).toMatchSnapshot();
      });
    });

    describe('[options]', () => {
      it('handles primitives', () => {
        expect(driver.formatFile({
          options: {
            emoji: true,
            'esproposal.class_instance_fields': 'ignore',
            traces: 3,
          },
        })).toMatchSnapshot();
      });

      it('handles arrays', () => {
        expect(driver.formatFile({
          options: {
            'module.file_ext': ['.js', '.jsx'],
          },
        })).toMatchSnapshot();
      });

      it('handles objects', () => {
        expect(driver.formatFile({
          options: {
            'module.name_mapper': {
              '^image![a-zA-Z0-9$_]+$': 'ImageStub',
            },
            'module.name_mapper.extension': {
              css: '<PROJECT_ROOT>/CSSFlowStub.js.flow',
            },
          },
        })).toMatchSnapshot();
      });

      it('handles regex', () => {
        expect(driver.formatFile({
          options: {
            suppress_comment: /(.|\n)*\$FlowFixMe/,
          },
        })).toMatchSnapshot();
      });

      it('handles escaped string', () => {
        expect(driver.formatFile({
          options: {
            suppress_comment: '\\\\(.\\\\|\\n\\\\)*\\\\$FlowFixMe',
          },
        })).toMatchSnapshot();
      });
    });

    describe('[version]', () => {
      it('supports', () => {
        expect(driver.formatFile({
          version: '1.2.3',
        })).toMatchSnapshot();
      });

      it('handles empty', () => {
        expect(driver.formatFile({
          version: '',
        })).toMatchSnapshot();
      });
    });
  });
});
