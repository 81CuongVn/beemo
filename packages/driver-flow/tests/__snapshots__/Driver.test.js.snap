// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlowDriver formatFile() [ignore] handles empty 1`] = `
"[ignore]
"
`;

exports[`FlowDriver formatFile() [ignore] supports 1`] = `
"[ignore]
.*/__tests__/.*
"
`;

exports[`FlowDriver formatFile() [include] handles empty 1`] = `
"[include]
"
`;

exports[`FlowDriver formatFile() [include] supports 1`] = `
"[include]
./foo.js
./bar
"
`;

exports[`FlowDriver formatFile() [libs] handles empty 1`] = `
"[libs]
"
`;

exports[`FlowDriver formatFile() [libs] supports 1`] = `
"[libs]
./foo.js.flow
"
`;

exports[`FlowDriver formatFile() [lints] handles values as numbers 1`] = `
"[lints]
all=error
untyped-type-import=warn
sketchy-null-bool=off
"
`;

exports[`FlowDriver formatFile() [lints] supports 1`] = `
"[lints]
all=error
untyped-type-import=warn
sketchy-null-bool=off
"
`;

exports[`FlowDriver formatFile() [lints] supports dashed keys 1`] = `
"[lints]
all=error
untyped-type-import=warn
sketchy-null-bool=off
"
`;

exports[`FlowDriver formatFile() [options] handles arrays 1`] = `
"[options]
module.file_ext=.js
module.file_ext=.jsx
"
`;

exports[`FlowDriver formatFile() [options] handles escaped string 1`] = `
"[options]
suppress_comment=\\\\\\\\(.\\\\\\\\|\\\\n\\\\\\\\)*\\\\\\\\$FlowFixMe
"
`;

exports[`FlowDriver formatFile() [options] handles objects 1`] = `
"[options]
module.name_mapper='^image![a-zA-Z0-9$_]+$' -> 'ImageStub'
module.name_mapper.extension='css' -> '<PROJECT_ROOT>/CSSFlowStub.js.flow'
"
`;

exports[`FlowDriver formatFile() [options] handles primitives 1`] = `
"[options]
emoji=true
esproposal.class_instance_fields=ignore
traces=3
"
`;

exports[`FlowDriver formatFile() [options] handles regex 1`] = `
"[options]
suppress_comment=\\\\(.\\\\|\\\\n\\\\)*\\\\$FlowFixMe
"
`;

exports[`FlowDriver formatFile() [version] handles empty 1`] = `""`;

exports[`FlowDriver formatFile() [version] supports 1`] = `
"[version]
1.2.3
"
`;
